lite/provider_test.go

```diff
@@ -1,149 +1,140 @@
-	fc, err = p.GetByHeight(5000)
+	// Make sure we get the last hash if we overstep.
+	fc, err = p.LatestFullCommit(chainID, 1, 5000)
 	if assert.Nil(err) {
-		assert.Equal(commits[count-1].Height(), fc.Height())
-		assert.Equal(commits[count-1], fc)
+		assert.Equal(fcz[count-1].Height(), fc.Height())
+		assert.Equal(fcz[count-1], fc)
 	}
 
-	// and middle ones as well
-	fc, err = p.GetByHeight(47)
+	// ... and middle ones as well.
+	fc, err = p.LatestFullCommit(chainID, 1, 47)
 	if assert.Nil(err) {
 		// we only step by 10, so 40 must be the one below this
 		assert.EqualValues(40, fc.Height())
 	}
 
 }
 
-// this will make a get height, and if it is good, set the data as well
-func checkGetHeight(t *testing.T, p lite.Provider, ask, expect int64) {
-	fc, err := p.GetByHeight(ask)
-	require.Nil(t, err, "GetByHeight")
+// This will make a get height, and if it is good, set the data as well.
+func checkLatestFullCommit(t *testing.T, p PersistentProvider, chainID string, ask, expect int64) {
+	fc, err := p.LatestFullCommit(chainID, 1, ask)
+	require.Nil(t, err)
 	if assert.Equal(t, expect, fc.Height()) {
-		err = p.StoreCommit(fc)
-		require.Nil(t, err, "StoreCommit")
+		err = p.SaveFullCommit(fc)
+		require.Nil(t, err)
 	}
 }
 
-func TestCacheGetsBestHeight(t *testing.T) {
-	// assert, require := assert.New(t), require.New(t)
+func TestMultiLatestFullCommit(t *testing.T) {
 	require := require.New(t)
 
-	// we will write data to the second level of the cache (p2),
-	// and see what gets cached, stored in
-	p := lite.NewMemStoreProvider()
-	p2 := lite.NewMemStoreProvider()
-	cp := lite.NewCacheProvider(p, p2)
+	// We will write data to the second level of the cache (p2), and see what
+	// gets cached/stored in.
+	p := NewDBProvider("mem1", dbm.NewMemDB())
+	p2 := NewDBProvider("mem2", dbm.NewMemDB())
+	cp := NewMultiProvider(p, p2)
 
 	chainID := "cache-best-height"
 	appHash := []byte("01234567")
-	keys := lite.GenValKeys(5)
+	keys := genPrivKeys(5)
 	count := 10
 
-	// set a bunch of commits
+	// Set a bunch of full commits.
 	for i := 0; i < count; i++ {
 		vals := keys.ToValidators(10, int64(count/2))
 		h := int64(10 * (i + 1))
-		fc := keys.GenFullCommit(chainID, h, nil, vals, appHash, []byte("params"), []byte("results"), 0, 5)
-		err := p2.StoreCommit(fc)
+		fc := keys.GenFullCommit(chainID, h, nil, vals, vals, appHash, []byte("params"), []byte("results"), 0, 5)
+		err := p2.SaveFullCommit(fc)
 		require.NoError(err)
 	}
 
-	// let's get a few heights from the cache and set them proper
-	checkGetHeight(t, cp, 57, 50)
-	checkGetHeight(t, cp, 33, 30)
+	// Get a few heights from the cache and set them proper.
+	checkLatestFullCommit(t, cp, chainID, 57, 50)
+	checkLatestFullCommit(t, cp, chainID, 33, 30)
 
 	// make sure they are set in p as well (but nothing else)
-	checkGetHeight(t, p, 44, 30)
-	checkGetHeight(t, p, 50, 50)
-	checkGetHeight(t, p, 99, 50)
+	checkLatestFullCommit(t, p, chainID, 44, 30)
```

Comment:
Test is not very clear. I guess idea is to first have empty cache
and then to write some commits to a cache and confirms that multi-provider
returns the same, but it's not obvious from the test.

Reply:
TODO make issue